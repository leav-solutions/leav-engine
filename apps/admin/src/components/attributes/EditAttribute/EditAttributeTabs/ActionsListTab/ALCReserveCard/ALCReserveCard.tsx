// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import React, {useEffect, useRef, useState} from 'react';
import {useDrag} from 'react-dnd';
import {getEmptyImage} from 'react-dnd-html5-backend';
import {useTranslation} from 'react-i18next';
import {Button, Card, Icon} from 'semantic-ui-react';
import TypeTag from '../ALCTypeTag';
import {IAction, IColorDic, IParamInput} from '../interfaces/interfaces';
import itemTypes from '../ItemTypes';

interface IALCReserveCardProps {
    id: string;
    action: IAction;
    moveCard?: (id: string, isOver: boolean, to: number) => void;
    findCard?: (id: string) => number | undefined;
    origin: string;
    addActionToList: (actionName: string, atIndex: number) => void;
    removeActionFromList?: (id: string) => void;
    marginTop?: string;
    currentIndex?: number;
    setCurrentIndex?: (idx: number) => void;
    connectionState?: {top: boolean; bottom: boolean};
    getConnectorStatus?: (indx: number, action: {}) => boolean | undefined;
    colorTypeDictionnary: IColorDic;
    changeParam?: (input: IParamInput) => void;
    index?: number;
    dragging?: boolean;
}

function ALCReserveCard({
    id,
    action,
    origin,
    addActionToList,
    removeActionFromList,
    setCurrentIndex,
    connectionState,
    colorTypeDictionnary
}: IALCReserveCardProps): JSX.Element {
    const {t} = useTranslation();
    const container = useRef(null);
    const [typesOpen, toggleTypes] = useState(false);
    const [internalWidth, setWidth] = useState(null);

    const [, drag, preview] = useDrag({
        item: {
            type: itemTypes.ACTION,
            id,
            originalIndex: -1,
            origin,
            action,
            connectionState,
            colorTypeDictionnary,
            width: internalWidth
        },
        end(item: any, monitor) {
            if (setCurrentIndex) {
                setCurrentIndex(-1);
            }
            if (!monitor.didDrop()) {
                if (removeActionFromList && !item.action.isSystem) {
                    removeActionFromList(id);
                }
            }
        },
        collect: monitor => ({
            isDragging: !!monitor.isDragging()
        })
    });

    useEffect(() => {
        preview(getEmptyImage(), {captureDraggingState: false});
        // @ts-ignore
        setWidth(container && container.current && container.current.offsetWidth);
    }, [preview]);

    const handleToggleTypes = () => {
        toggleTypes(!typesOpen);
    };

    const onAddButtonClicked = () => {
        tryAndAdd(action.id, -1);
    };

    const tryAndAdd = (actId: string, atIndex: number) => {
        if (addActionToList) {
            addActionToList(actId, atIndex);
        }
    };

    const inputs = action.input_types;
    const outputs = action.output_types;

    return (
        <div ref={container}>
            <div ref={node => drag(node)} style={{paddingBottom: '5px'}}>
                <Card fluid>
                    <Card.Content>
                        <Card.Header>{action.name}</Card.Header>
                        <Card.Description>{action.description}</Card.Description>
                        <Button
                            style={{
                                position: 'absolute',
                                right: '2px',
                                top: '5px',
                                fontSize: '0.8em'
                            }}
                            circular
                            icon="add"
                            onClick={onAddButtonClicked}
                        />
                        <div
                            style={{
                                textAlign: 'right'
                            }}
                            onClick={handleToggleTypes}
                        >
                            <Card.Meta>
                                {typesOpen ? t('attributes.hide_types') : t('attributes.display_types')}
                                <Icon name={typesOpen ? 'triangle down' : 'triangle right'} />
                            </Card.Meta>
                        </div>
                    </Card.Content>
                    {typesOpen && (
                        <Card.Content extra>
                            <div style={{padding: 0}}>
                                <Icon name="triangle right" />
                                <span>{t('attributes.input_types')}: </span>
                                {inputs &&
                                    colorTypeDictionnary[inputs[0]] &&
                                    inputs.map((input, i) => (
                                        <TypeTag key={i} color={colorTypeDictionnary[input]} input={input} />
                                    ))}
                            </div>
                            <div style={{marginTop: '5px'}}>
                                <Icon name="triangle left" />
                                <span>{t('attributes.output_types')}: </span>
                                {outputs &&
                                    colorTypeDictionnary[outputs[0]] &&
                                    outputs.map((output, i) => (
                                        <TypeTag key={i} color={colorTypeDictionnary[output]} input={output} />
                                    ))}
                            </div>
                        </Card.Content>
                    )}
                </Card>
            </div>
        </div>
    );
}

export default ALCReserveCard;
