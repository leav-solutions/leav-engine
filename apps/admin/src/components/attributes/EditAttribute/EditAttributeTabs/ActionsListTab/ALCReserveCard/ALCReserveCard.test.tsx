// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {render} from 'enzyme';
import React from 'react';
import {DndProvider} from 'react-dnd';
import {TestBackend} from 'react-dnd-test-backend';
import ALCReserveCard from './ALCReserveCard';

function placeholder() {
    return undefined;
}

jest.mock('react-dnd', () => ({
    useDrag: () => {
        const isDragging = false;
        const drag = () => {
            return true;
        };
        const preview = () => {
            return true;
        };
        return [{isDragging}, drag, preview];
    },
    useDrop: sth => {
        const isOver = false;
        const drop = () => {
            return true;
        };
        return [{isOver}, drop];
    },
    DndProvider: ({backend, children}) => children
}));

describe('ALCReserveCard', () => {
    test('Snapshot test', async () => {
        const comp = render(
            <DndProvider backend={TestBackend}>
                <ALCReserveCard
                    id={'0'}
                    action={{
                        id: 'action',
                        list_id: 0,
                        name: 'action',
                        description: 'action',
                        input_types: [],
                        output_types: [],
                        params: [],
                        isSystem: false
                    }}
                    origin="ALCList"
                    addActionToList={placeholder}
                    colorTypeDictionnary={{int: []}}
                />
            </DndProvider>
        );

        expect(comp).toMatchSnapshot();
    });
});
