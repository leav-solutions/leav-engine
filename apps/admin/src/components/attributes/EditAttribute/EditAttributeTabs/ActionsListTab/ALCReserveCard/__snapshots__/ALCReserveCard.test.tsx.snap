// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ALCReserveCard Snapshot test 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": {},
    "children": [
      Node {
        "attribs": {
          "style": "padding-bottom:5px",
        },
        "children": [
          Node {
            "attribs": {
              "class": "ui fluid card",
            },
            "children": [
              Node {
                "attribs": {
                  "class": "content",
                },
                "children": [
                  Node {
                    "attribs": {
                      "class": "header",
                    },
                    "children": [
                      Node {
                        "data": "action",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Node {
                      "attribs": {
                        "class": "description",
                      },
                      "children": [
                        Node {
                          "data": "action",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Node {
                        "attribs": {
                          "class": "ui circular icon button",
                          "style": "position:absolute;right:2px;top:5px;font-size:0.8em",
                        },
                        "children": [
                          Node {
                            "attribs": {
                              "aria-hidden": "true",
                              "class": "add icon",
                            },
                            "children": [],
                            "name": "i",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": {
                              "aria-hidden": undefined,
                              "class": undefined,
                            },
                            "x-attribsPrefix": {
                              "aria-hidden": undefined,
                              "class": undefined,
                            },
                          },
                        ],
                        "name": "button",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": Node {
                          "attribs": {
                            "style": "text-align:right",
                          },
                          "children": [
                            Node {
                              "attribs": {
                                "class": "meta",
                              },
                              "children": [
                                Node {
                                  "data": "attributes.display_types",
                                  "next": Node {
                                    "attribs": {
                                      "aria-hidden": "true",
                                      "class": "triangle right icon",
                                    },
                                    "children": [],
                                    "name": "i",
                                    "namespace": "http://www.w3.org/1999/xhtml",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": [Circular],
                                    "type": "tag",
                                    "x-attribsNamespace": {
                                      "aria-hidden": undefined,
                                      "class": undefined,
                                    },
                                    "x-attribsPrefix": {
                                      "aria-hidden": undefined,
                                      "class": undefined,
                                    },
                                  },
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "text",
                                },
                                Node {
                                  "attribs": {
                                    "aria-hidden": "true",
                                    "class": "triangle right icon",
                                  },
                                  "children": [],
                                  "name": "i",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": Node {
                                    "data": "attributes.display_types",
                                    "next": [Circular],
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "text",
                                  },
                                  "type": "tag",
                                  "x-attribsNamespace": {
                                    "aria-hidden": undefined,
                                    "class": undefined,
                                  },
                                  "x-attribsPrefix": {
                                    "aria-hidden": undefined,
                                    "class": undefined,
                                  },
                                },
                              ],
                              "name": "div",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": {
                                "class": undefined,
                              },
                              "x-attribsPrefix": {
                                "class": undefined,
                              },
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": [Circular],
                          "type": "tag",
                          "x-attribsNamespace": {
                            "style": undefined,
                          },
                          "x-attribsPrefix": {
                            "style": undefined,
                          },
                        },
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": {
                          "class": undefined,
                          "style": undefined,
                        },
                        "x-attribsPrefix": {
                          "class": undefined,
                          "style": undefined,
                        },
                      },
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": {
                        "class": undefined,
                      },
                      "x-attribsPrefix": {
                        "class": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": {
                      "class": undefined,
                    },
                    "x-attribsPrefix": {
                      "class": undefined,
                    },
                  },
                  Node {
                    "attribs": {
                      "class": "description",
                    },
                    "children": [
                      Node {
                        "data": "action",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Node {
                      "attribs": {
                        "class": "ui circular icon button",
                        "style": "position:absolute;right:2px;top:5px;font-size:0.8em",
                      },
                      "children": [
                        Node {
                          "attribs": {
                            "aria-hidden": "true",
                            "class": "add icon",
                          },
                          "children": [],
                          "name": "i",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": {
                            "aria-hidden": undefined,
                            "class": undefined,
                          },
                          "x-attribsPrefix": {
                            "aria-hidden": undefined,
                            "class": undefined,
                          },
                        },
                      ],
                      "name": "button",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Node {
                        "attribs": {
                          "style": "text-align:right",
                        },
                        "children": [
                          Node {
                            "attribs": {
                              "class": "meta",
                            },
                            "children": [
                              Node {
                                "data": "attributes.display_types",
                                "next": Node {
                                  "attribs": {
                                    "aria-hidden": "true",
                                    "class": "triangle right icon",
                                  },
                                  "children": [],
                                  "name": "i",
                                  "namespace": "http://www.w3.org/1999/xhtml",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": [Circular],
                                  "type": "tag",
                                  "x-attribsNamespace": {
                                    "aria-hidden": undefined,
                                    "class": undefined,
                                  },
                                  "x-attribsPrefix": {
                                    "aria-hidden": undefined,
                                    "class": undefined,
                                  },
                                },
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                              Node {
                                "attribs": {
                                  "aria-hidden": "true",
                                  "class": "triangle right icon",
                                },
                                "children": [],
                                "name": "i",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": null,
                                "parent": [Circular],
                                "prev": Node {
                                  "data": "attributes.display_types",
                                  "next": [Circular],
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "text",
                                },
                                "type": "tag",
                                "x-attribsNamespace": {
                                  "aria-hidden": undefined,
                                  "class": undefined,
                                },
                                "x-attribsPrefix": {
                                  "aria-hidden": undefined,
                                  "class": undefined,
                                },
                              },
                            ],
                            "name": "div",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": {
                              "class": undefined,
                            },
                            "x-attribsPrefix": {
                              "class": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": {
                          "style": undefined,
                        },
                        "x-attribsPrefix": {
                          "style": undefined,
                        },
                      },
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": {
                        "class": undefined,
                        "style": undefined,
                      },
                      "x-attribsPrefix": {
                        "class": undefined,
                        "style": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": Node {
                      "attribs": {
                        "class": "header",
                      },
                      "children": [
                        Node {
                          "data": "action",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": {
                        "class": undefined,
                      },
                      "x-attribsPrefix": {
                        "class": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": {
                      "class": undefined,
                    },
                    "x-attribsPrefix": {
                      "class": undefined,
                    },
                  },
                  Node {
                    "attribs": {
                      "class": "ui circular icon button",
                      "style": "position:absolute;right:2px;top:5px;font-size:0.8em",
                    },
                    "children": [
                      Node {
                        "attribs": {
                          "aria-hidden": "true",
                          "class": "add icon",
                        },
                        "children": [],
                        "name": "i",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": {
                          "aria-hidden": undefined,
                          "class": undefined,
                        },
                        "x-attribsPrefix": {
                          "aria-hidden": undefined,
                          "class": undefined,
                        },
                      },
                    ],
                    "name": "button",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Node {
                      "attribs": {
                        "style": "text-align:right",
                      },
                      "children": [
                        Node {
                          "attribs": {
                            "class": "meta",
                          },
                          "children": [
                            Node {
                              "data": "attributes.display_types",
                              "next": Node {
                                "attribs": {
                                  "aria-hidden": "true",
                                  "class": "triangle right icon",
                                },
                                "children": [],
                                "name": "i",
                                "namespace": "http://www.w3.org/1999/xhtml",
                                "next": null,
                                "parent": [Circular],
                                "prev": [Circular],
                                "type": "tag",
                                "x-attribsNamespace": {
                                  "aria-hidden": undefined,
                                  "class": undefined,
                                },
                                "x-attribsPrefix": {
                                  "aria-hidden": undefined,
                                  "class": undefined,
                                },
                              },
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                            Node {
                              "attribs": {
                                "aria-hidden": "true",
                                "class": "triangle right icon",
                              },
                              "children": [],
                              "name": "i",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": null,
                              "parent": [Circular],
                              "prev": Node {
                                "data": "attributes.display_types",
                                "next": [Circular],
                                "parent": [Circular],
                                "prev": null,
                                "type": "text",
                              },
                              "type": "tag",
                              "x-attribsNamespace": {
                                "aria-hidden": undefined,
                                "class": undefined,
                              },
                              "x-attribsPrefix": {
                                "aria-hidden": undefined,
                                "class": undefined,
                              },
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": {
                            "class": undefined,
                          },
                          "x-attribsPrefix": {
                            "class": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": {
                        "style": undefined,
                      },
                      "x-attribsPrefix": {
                        "style": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": Node {
                      "attribs": {
                        "class": "description",
                      },
                      "children": [
                        Node {
                          "data": "action",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": Node {
                        "attribs": {
                          "class": "header",
                        },
                        "children": [
                          Node {
                            "data": "action",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": {
                          "class": undefined,
                        },
                        "x-attribsPrefix": {
                          "class": undefined,
                        },
                      },
                      "type": "tag",
                      "x-attribsNamespace": {
                        "class": undefined,
                      },
                      "x-attribsPrefix": {
                        "class": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": {
                      "class": undefined,
                      "style": undefined,
                    },
                    "x-attribsPrefix": {
                      "class": undefined,
                      "style": undefined,
                    },
                  },
                  Node {
                    "attribs": {
                      "style": "text-align:right",
                    },
                    "children": [
                      Node {
                        "attribs": {
                          "class": "meta",
                        },
                        "children": [
                          Node {
                            "data": "attributes.display_types",
                            "next": Node {
                              "attribs": {
                                "aria-hidden": "true",
                                "class": "triangle right icon",
                              },
                              "children": [],
                              "name": "i",
                              "namespace": "http://www.w3.org/1999/xhtml",
                              "next": null,
                              "parent": [Circular],
                              "prev": [Circular],
                              "type": "tag",
                              "x-attribsNamespace": {
                                "aria-hidden": undefined,
                                "class": undefined,
                              },
                              "x-attribsPrefix": {
                                "aria-hidden": undefined,
                                "class": undefined,
                              },
                            },
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                          Node {
                            "attribs": {
                              "aria-hidden": "true",
                              "class": "triangle right icon",
                            },
                            "children": [],
                            "name": "i",
                            "namespace": "http://www.w3.org/1999/xhtml",
                            "next": null,
                            "parent": [Circular],
                            "prev": Node {
                              "data": "attributes.display_types",
                              "next": [Circular],
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                            "type": "tag",
                            "x-attribsNamespace": {
                              "aria-hidden": undefined,
                              "class": undefined,
                            },
                            "x-attribsPrefix": {
                              "aria-hidden": undefined,
                              "class": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": {
                          "class": undefined,
                        },
                        "x-attribsPrefix": {
                          "class": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Node {
                      "attribs": {
                        "class": "ui circular icon button",
                        "style": "position:absolute;right:2px;top:5px;font-size:0.8em",
                      },
                      "children": [
                        Node {
                          "attribs": {
                            "aria-hidden": "true",
                            "class": "add icon",
                          },
                          "children": [],
                          "name": "i",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": {
                            "aria-hidden": undefined,
                            "class": undefined,
                          },
                          "x-attribsPrefix": {
                            "aria-hidden": undefined,
                            "class": undefined,
                          },
                        },
                      ],
                      "name": "button",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": Node {
                        "attribs": {
                          "class": "description",
                        },
                        "children": [
                          Node {
                            "data": "action",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "text",
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": Node {
                          "attribs": {
                            "class": "header",
                          },
                          "children": [
                            Node {
                              "data": "action",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "text",
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": [Circular],
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": {
                            "class": undefined,
                          },
                          "x-attribsPrefix": {
                            "class": undefined,
                          },
                        },
                        "type": "tag",
                        "x-attribsNamespace": {
                          "class": undefined,
                        },
                        "x-attribsPrefix": {
                          "class": undefined,
                        },
                      },
                      "type": "tag",
                      "x-attribsNamespace": {
                        "class": undefined,
                        "style": undefined,
                      },
                      "x-attribsPrefix": {
                        "class": undefined,
                        "style": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": {
                      "style": undefined,
                    },
                    "x-attribsPrefix": {
                      "style": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": {
                  "class": undefined,
                },
                "x-attribsPrefix": {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": {
              "class": undefined,
            },
            "x-attribsPrefix": {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": {
          "style": undefined,
        },
        "x-attribsPrefix": {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": {},
    "x-attribsPrefix": {},
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": [
        Node {
          "attribs": {},
          "children": [
            Node {
              "attribs": {},
              "children": [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": {},
                "children": [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": {},
                "x-attribsPrefix": {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": {},
              "x-attribsPrefix": {},
            },
            Node {
              "attribs": {},
              "children": [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": {},
                "children": [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": {},
                "x-attribsPrefix": {},
              },
              "type": "tag",
              "x-attribsNamespace": {},
              "x-attribsPrefix": {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": {},
          "x-attribsPrefix": {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
